osdconfig/api.go

```diff
@@ -0,0 +1,275 @@
+	if err != nil {
+		return nil, err
+	}
+
+	if kvPair == nil {
+		return nil, errors.New("no data fetched from kvdb store")
+	}
+
+	config := new(ClusterConfig)
+	if err := json.Unmarshal(kvPair.Value, config); err != nil {
+		return nil, err
+	}
+
+	return config, nil
+}
+
+// SetClusterConf sets cluster config in kvdb
+func (manager *configManager) SetClusterConf(config *ClusterConfig) error {
+	if config == nil {
+		return errors.New("input is nil")
+	}
+
+	// push into kvdb
+	_, err := manager.cc.Put(filepath.Join(base_key, cluster_key), config, 0)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// GetNodeConf retrieves node config data using nodeID
+func (manager *configManager) GetNodeConf(nodeID string) (*NodeConfig, error) {
+	if len(nodeID) == 0 {
+		return nil, errors.New("node id input is empty")
+	}
+
+	// get json from kvdb and unmarshal into config
+	kvPair, err := manager.cc.Get(getNodeKeyFromNodeID(nodeID))
+	if err != nil {
+		return nil, err
+	}
+
+	if kvPair == nil {
+		return nil, errors.New("no data fetched from kvdb store")
+	}
+
+	config := new(NodeConfig)
+	if err := json.Unmarshal(kvPair.Value, config); err != nil {
+		return nil, err
+	}
+
+	return config, nil
+}
+
+// SetNodeConf sets node config data in kvdb
+func (manager *configManager) SetNodeConf(config *NodeConfig) error {
+	if config == nil {
+		return errors.New("input is nil")
+	}
+
+	if len(config.NodeId) == 0 {
+		return errors.New("node id in input is empty")
+	}
+
+	// push into kvdb
+	_, err := manager.cc.Put(getNodeKeyFromNodeID(config.NodeId), config, 0)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// RegisterCB registers callback functions
+// callback to be registered is expected to be a service delivering a channel to write on
+func (manager *configManager) RegisterCB(name string, opt interface{},
+	cb func(ctx context.Context,
+		opt interface{}) (chan<- *DataToCallback, <-chan *DataFromCallback)) error {
+
+	// obtain lock since registering a callback changes the state of manager object
+	manager.Lock()
+	defer manager.Unlock()
+
+	// callbacks are registered by the "name" string. only one per name
+	if _, present := manager.cb[name]; !present {
+		cbi := new(callbackData)
+		cbi.f = cb
+		cbi.opt = opt
+		manager.cb[name] = cbi
+	} else {
+		return errors.New("callback already registered with this name: " + name)
+	}
+
+	return nil
+}
+
+// Run loops over all registered callbacks and executes them.
+// It returns a channel on which you only read and it contains the errors.
+func (manager *configManager) Run(wd *DataToCallback) <-chan error {
+	manager.Lock()
+	// re-establish the context
+	// derive local contexts from parent context
+	manager.localCtx, manager.localCancel = context.WithCancel(manager.ctx)
+	manager.Unlock()
+
+	// declare a channel on which callers of Run can wait on
+	wait := make(chan error)
+	go func(waitChan chan error, dataToCallback *DataToCallback) {
```

Comment:
- i think the channel can go 
- prefer running callbacks in order, since there are no guarantees on running them via go routines - so no advantage really

Reply:
wait channel is gone now. It was possible to work directly with the contexts.